analyzers/tests/SonarAnalyzer.UnitTest/SymbolicExecution/Roslyn/RoslynSymbolicExecutionTest.Operations.cs

```diff
@@ -140,6 +140,27 @@ public void SimpleAssignment_ToUnsupported_FromLiteral(string snippet)
             validator.ValidateTag("Target", x => (x?.HasConstraint(DummyConstraint.Dummy) ?? false).Should().BeFalse());
         }
 
+        [TestMethod]
+        public void SimpleAssignment_ResetsObjectConstraint()
+        {
+            var validator = SETestContext.CreateCS("""
+                if (arg == null)
+                {
+                    Tag("BeforeReassignment");
+                    arg = Guid.NewGuid().ToString("N");
+                    Tag("AfterReassignment");
+                }
+                Tag("End");
+                arg.ToString();
+                """, ", string arg").Validator;
+            var arg = validator.Symbol("arg");
+            validator.TagStates("BeforeReassignment").Should().ContainSingle().Which[arg].Should().HaveOnlyConstraint(ObjectConstraint.Null);
+            validator.TagStates("AfterReassignment").Should().ContainSingle().Which[arg].Should().HaveNoConstraints();
+            validator.TagStates("End").Should().SatisfyRespectively(
+                x => x[arg].Should().HaveOnlyConstraint(ObjectConstraint.NotNull),
+                x => x[arg].Should().HaveNoConstraints());
```

Comment:
@pavel-mikula-sonarsource This state is the root cause for the FP:

S3900 is getting triggered because `arg` has no object constraint here.

I see two options for how to solve this in `PublicMethodArgumentsShouldBeCheckedForNull`
1. Use a "VariableReassigned" SymbolicValue which gets added on each re-assignment to the symbol. We can check for this in the "End" state.
2. Track reassigned parameters in a HashSet field in `PublicMethodArgumentsShouldBeCheckedForNull` and only raise if the de-referenced parameter without ObjectConstraint is not in this set.

Do you think any of the two options would work and if so which one would you prefer?